internal/users/tempentries/tempentries.go

```diff
@@ -0,0 +1,151 @@
+		preAuthUserRecords:    newPreAuthUserRecords(idGenerator),
+		temporaryGroupRecords: newTemporaryGroupRecords(idGenerator),
+	}
+}
+
+// UserByID returns the user information for the given user ID.
+func (r *TemporaryRecords) UserByID(uid uint32) (types.UserEntry, error) {
+	user, err := r.temporaryUserRecords.userByID(uid)
+	if errors.Is(err, NoDataFoundError{}) {
+		user, err = r.preAuthUserRecords.userByID(uid)
+	}
+	return user, err
+}
+
+// UserByName returns the user information for the given user name.
+func (r *TemporaryRecords) UserByName(name string) (types.UserEntry, error) {
+	user, err := r.temporaryUserRecords.userByName(name)
+	if errors.Is(err, NoDataFoundError{}) {
+		user, err = r.preAuthUserRecords.userByName(name)
+	}
+	return user, err
+}
+
+// RegisterUser registers a temporary user with a unique UID in our NSS handler (in memory, not in the database).
+//
+// Returns the generated UID and a cleanup function that should be called to remove the temporary user once the user was
+// added to the database.
+func (r *TemporaryRecords) RegisterUser(name string) (uid uint32, cleanup func() error, err error) {
+	r.temporaryUserRecords.registerMutex.Lock()
+	defer r.temporaryUserRecords.registerMutex.Unlock()
+
+	// Check if there is a temporary  user with the same login name.
+	_, err = r.temporaryUserRecords.userByName(name)
+	if err != nil && !errors.Is(err, NoDataFoundError{}) {
+		return 0, nil, fmt.Errorf("could not check if temporary user %q already exists: %w", name, err)
+	}
+	if err == nil {
+		return 0, nil, fmt.Errorf("user %q already exists", name)
+	}
+
+	// Check if there is a pre-auth user with the same login name.
+	user, err := r.preAuthUserRecords.userByLogin(name)
+	if err != nil && !errors.Is(err, NoDataFoundError{}) {
+		return 0, nil, fmt.Errorf("could not check if pre-auth user %q already exists: %w", name, err)
+	}
+	if err == nil {
+		// There is a pre-auth user with the same login name. Now that the user authenticated successfully, we can
+		// replace the pre-auth user with a temporary user.
+		var tmpID string
+		tmpID, cleanup, err = r.addTemporaryUser(user.UID, name)
+		if err != nil {
+			return 0, nil, fmt.Errorf("could not add temporary user record: %w", err)
+		}
+
+		// Remove the pre-auth user from the pre-auth user records.
+		r.deletePreAuthUser(user.UID)
+
+		// Check if the UID and name are unique.
+		unique, err := r.temporaryUserRecords.uniqueNameAndUID(name, user.UID, tmpID)
+		if err != nil {
+			err = fmt.Errorf("checking UID and name uniqueness: %w", err)
+			if cleanupErr := cleanup(); cleanupErr != nil {
+				err = errors.Join(err, cleanupErr)
+			}
+			return 0, nil, err
+		}
+		if !unique {
+			err = fmt.Errorf("UID (%d) or name (%q) from pre-auth user are not unique", user.UID, name)
+			if cleanupErr := cleanup(); cleanupErr != nil {
+				err = errors.Join(err, cleanupErr)
+			}
+			return 0, nil, err
+		}
+
+		return user.UID, cleanup, nil
+	}
```

Comment:
Maybe move this case to another function?

Reply:
done